### PR TITLE
Add raising method variants `Enumerable#find!` and `#index!`

### LINKED ISSUE #11565: `Enumerable#find` and `Enumerable#index` don't have raising variants.
## Feature Request
- Is your feature request related to a problem? Please describe clearly and concisely what is it.

Both `Enumerable#find` or `Enumerable#index` will return nil if the search for element isn't found. In some cases there is already knowledge that there will be an element that matches the block. In that case the user will be forced to do something like
```
array.find { .. }.not_nil!
array.index { .. }.not_nil!
```
or equivalent to erase the nilability of the return type. 

- Describe the feature you would like, optionally illustrated by examples, and how it will solve the above problem.

I propose adding 
```
array.find! { .. }
array.index! { .. }
```
to Enumerable, where if the block doesn't return true it will raise an exception. I'd suggest adding a `Enumerable#NotFoundError` for these, but am open to other suggestions.

- Describe considered alternative solutions, and the reasons why you have not proposed them as a solution here.

Another common pattern for doing the same thing would be to separate the existing methods into `find` and `find?`, `index` and `index?` respectively. That would also be fine and match existing patterns for this, but as the methods already exist I'd prefer to not break backwards compatibility.

- Does it break backward compatibility, if yes then what's the migration path?

No

### PR DESCRIPTION
**Pull request by yxhuvud**
Adds `Enumerable#find!` and `Enumerable#index!`

Resolves #11565

### CODE REVIEW

File: src/enumerable.cr [hunk 1]
```diff
@@ -642,6 +662,31 @@ module Enumerable(T)
...
+  # ["Alice", "Bob"].index! { |name| name.size < 4 } # => 1 (Bob's index)
+  # ```
+  #
+  # Raises `Enumerable::NotFoundError` if there are no element for which the block is `true`.
+  def index!(& : T ->) : Int32
+    each_with_index do |e, i|
+      return i if yield e
+    end
+    raise Enumerable::NotFoundError.new
+  end
```
Comment (HertzDevil) [inline src/enumerable.cr:677]: This is simply `index { |e| yield e } || raise ...`, because a valid index is always non-nil.

Comment (oprypin) [reply src/enumerable.cr:677]: That is true but I think I prefer the current way it's written.

If I'm looking at 
```crystal
    each_with_index do |e, i|
      return i if yield e
    end
    raise Enumerable::NotFoundError.new
```
It's much clearer that it's following a pattern from
```crystal
    each_with_index do |e, i|
      return i if yield e
    end
    nil
```
than from
```crystal
    index { |e| yield e } || raise Enumerable::NotFoundError.new
```

---

The pattern is *also* consistent with the pattern from `find!` where you can't do this. So let's not cherry-pick places where this happens to be possible.

---

In addition, as written, it avoids a roundtrip through `nil`.

**Review by oprypin — APPROVED** [review]

File: src/enumerable.cr (hunk 1 above)
Comment (asterite) [reply src/enumerable.cr:677]: I also prefer the more concise way 🤷

**Comment by Blacksmoke16** [comment]
A bit of a meta question, but what determines that a method should have a `!` version instead of raising by default with a `? version? Ofc we can't do it now, but for Crystal 2.0, what about making the current methods raise and add `find?`/`index?`. That be more inline with like `Hash#[]` and `Hash#[]?` at least.

**Comment by oprypin** [comment]
@Blacksmoke16 That is a good point and indeed maybe that's how it should've been, but to me it seems like too big and subtle of a breakage even for 2.0

**Comment by asterite** [comment]
My intuition is that it depends on what is the common expectation. When you index an array or a hash, it's very likely you expect the value to be there. With index and find, you want to find something but it's likely you might not find it. The language reminds you of this.

**Comment by j8r** [comment]
@Blacksmoke16 Generally `!` methods are not really to differentiate raising vs returning nil, but more to show that the alternate behavior require more care.

For example, `merge` and `merge!`, `sort` and `sort!`, this methods mutate the original object vs creating a copy.
There are also the `to_i!` methods which have no overflow check, thus are less safe.

Generally, the safe preferred default choice is to use the non `!` method - which is logical as most methods don't have alternate behaviors.

Here, probably a raising `find` is less safe than one returning nil, which will force us to handle a common case where the element is not found, as @asterite said.

File: src/enumerable.cr (hunk 1 above)
Comment (yxhuvud) [reply src/enumerable.cr]: I updated to the suggested version but I don't really care. Both sides have good arguments and the code is fine either way, IMO.

File: src/enumerable.cr [hunk 2]
```diff
@@ -481,6 +487,20 @@ module Enumerable(T)
     if_none
   end
 
+  # Returns the first element in the collection for which the passed block is `true`.
+  # Raises `Enumerable::NotFoundError` if there are no element for which the block is `true`.
```
Comment (straight-shoota) [inline src/enumerable.cr:491]:
```suggestion
  # Raises `Enumerable::NotFoundError` if there is no element for which the block is truthy.
```

File: src/enumerable.cr [hunk 3]
```diff
@@ -644,6 +664,28 @@ module Enumerable(T)
     index { |e| e == obj }
   end
 
+  # Returns the index of the first element for which the passed block returns `true`.
+  #
+  # ```
+  # ["Alice", "Bob"].index! { |name| name.size < 4 } # => 1 (Bob's index)
+  # ```
+  #
+  # Raises `Enumerable::NotFoundError` if there are no element for which the block is `true`.
```
Comment (straight-shoota) [inline src/enumerable.cr:673]:
```suggestion
  # Raises `Enumerable::NotFoundError` if there is no element for which the block is truthy.
```

File: src/enumerable.cr [hunk 4]
```diff
@@ -644,6 +664,28 @@ module Enumerable(T)
...
+  def index!(& : T ->) : Int32
+    index { |e| yield e } || raise Enumerable::NotFoundError.new
+  end
+
+  # Returns the index of the object *obj* in the collection.
```
Comment (straight-shoota) [inline src/enumerable.cr:678]: If multiple elements equal `obj`, only the first index is returned.

"the object": Unnecessary specification 
```suggestion
  # Returns the first index of *obj* in the collection.
```

File: src/enumerable.cr [hunk 5]
```diff
@@ -481,6 +487,20 @@ module Enumerable(T)
     if_none
   end
 
+  # Returns the first element in the collection for which the passed block is `true`.
```
Comment (straight-shoota) [inline src/enumerable.cr:490]:
```suggestion
  # Returns the first element in the collection for which the passed block is truthy.
```

File: src/enumerable.cr [hunk 6]
```diff
@@ -644,6 +664,28 @@ module Enumerable(T)
     index { |e| e == obj }
   end
 
+  # Returns the index of the first element for which the passed block returns `true`.
```
Comment (straight-shoota) [inline src/enumerable.cr:667]:
```suggestion
  # Returns the index of the first element for which the passed block returns truthy.
```

Comment (yxhuvud) [reply src/enumerable.cr:667]: Can `truthy` be used like that? Wouldn't that have to be `a truthy value` or something like that?

File: src/enumerable.cr [hunk 7]
```diff
@@ -644,6 +664,28 @@ module Enumerable(T)
...
+    index { |e| yield e } || raise Enumerable::NotFoundError.new
+  end
+
+  # Returns the index of the object *obj* in the collection.
+  #
+  # ```
+  # ["Alice", "Bob"].index!("Alice") # => 0
+  # ```
+  #
+  # Raises `Enumerable::NotFoundError` if there are no element for which the block is `true`.
```
[the remark this answers is not in the document]
Comment (yxhuvud) [reply src/enumerable.cr:684]: The nonbang version uses `Returns `nil` if *obj* is not in the collection.`, which also seems fine?

File: src/enumerable.cr (hunk 6 above)
Comment (straight-shoota) [reply src/enumerable.cr:667]: Maybe just use `is truthy`? It's more concise than any of the above. And a block technically doesn't return, so it's also more precise.

Comment (yxhuvud) [reply src/enumerable.cr:667]: Right, it will also match the terminology of `find`.

**Review by straight-shoota — APPROVED** [review]

**Review by oprypin — APPROVED** [review]

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**